src/client/driver/driver.js

```diff
@@ -426,6 +444,139 @@ export default class Driver extends serviceUtils.EventEmitter {
         childIframeDriverLink.sendConfirmationMessage(id);
     }
 
+    _getTargetWindowNotFoundResult () {
+        return Promise.resolve({
+            success: false,
+            errCode: TEST_RUN_ERRORS.targetWindowNotFoundError
+        });
+    }
+
+    _getChildWindowValidateResult (arr) {
+        const success = arr.find(item => item.result.success);
+
+        if (success)
+            return success.result;
+
+        let errItem = arr.find(item => item.result.errCode === TEST_RUN_ERRORS.cannotCloseWindowWithChildrenError);
+
+        if (!errItem)
+            errItem = arr.find(item => !!item.result.errCode);
+
+        return errItem ? { errCode: errItem.result.errCode } : void 0;
+    }
+
+    _handleWindowValidation (msg, wnd, windowFoundHandler, WindowValidationMessageCtor) {
+        let promiseResult = null;
+
+        if (msg.windowId === this.windowId)
+            promiseResult = windowFoundHandler();
+
+        else if (!this.childWindowDriverLinks.length) {
+            promiseResult = Promise.resolve({
+                success: false,
+                errCode: TEST_RUN_ERRORS.targetWindowNotFoundError
+            });
+        }
+
+        else {
+            promiseResult = Promise.all(this.childWindowDriverLinks.map(childWindowDriverLink => {
```

Comment:
All these Promise work don't read well. Could we refactor it in this way:
```ts
function _handleWindowValidation () {
    return _validateWindow()
        .sendConfimationMessage({ ... })
}

function _validateWindow () {
   if (window.id === windowId)
      return windowFoundHandler();
  
  if (!this.childWindowDriverLinks.length)
    return Promise.resolve({ WINDOW_NOT_FOUND_ERROR });

  const searchQueries = this.childWindowDriverLinks.map(() => { ... });

  return Promise.all(searchQueries)
     .then(() => this._getChildWindowValidateResult())
}
```

Reply:
And we can use the same pattern for `_handleWindowClose` & `_handleSwitchToWindow`. Or even use it like
```ts
function _handleRequest (msg, handler) {
  return handler().then(() => sendConfirmationMessage());
}

...
if(MSG.type===VALIDATE_WINDOW)
   this._handleWindowRequest(msg, (...args) => this._validateWindow(...args))
```

Reply:
Or
```ts
function _handleRequest (msg, handler) {
  return handler
    .call(this, msg)
    .then(() => sendConfirmationMessage());
}

...
if(MSG.type===VALIDATE_WINDOW)
   this._handleWindowRequest(msg, this._validateWindow)
```

Reply:
I think your first case is more readable. I would prefer the first on if you don't mind. Without additional extracting `handleRequest` and conditional method calls